### PR TITLE
Feat/ios/create gain node

### PR DESCRIPTION
**Pull request by hubgan**

### CODE REVIEW

File: ios/nodes/AudioNode/AudioNode.m [hunk 1]
```diff
@@ -1,29 +1,64 @@
...
+        return true;
+    }
+    
+    for (AudioNode *cn in node.connectedNodes) {
+        if ([self findNodesToDeprocessHelper:cn]) {
+            return true;
+        }
     }
+    
+    return false;
```
Comment (hubgan) [inline ios/nodes/AudioNode/AudioNode.m]: I think it's more clear than before. Right now when we will add new node, we don't have to worry about passing different pointers to this specific node and can focus on implementing core methods for that node

File: ios/nodes/GainNode/GainNode.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+#import "AudioNode.h"
+
+@interface GainNode : AudioNode
+
+@property (nonatomic, assign) float gain;
+
+- (instancetype)init:(AudioContext *)context;
+
+- (void)changeGain:(float)gain;
```
Comment (michalsek) [inline ios/nodes/GainNode/GainNode.h]: why not `setGain`?

File: cpp/AudioParam/ios/AudioParamWrapper.cpp [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#ifndef ANDROID
+#include "AudioParamWrapper.h"
+
+namespace audiocontext{
+
```
Comment (michalsek) [inline cpp/AudioParam/ios/AudioParamWrapper.cpp]:
```suggestion
namespace audiocontext
{

```

File: cpp/GainNode/ios/GainNodeWrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,14 @@
+#ifndef ANDROID
```
Comment (michalsek) [inline cpp/GainNode/ios/GainNodeWrapper.cpp:1]: Can't comment on empty file: can you remove the `.keep` file as it is not more necessary?

File: cpp/StereoPannerNode/ios/StereoPannerNodeWrapper.cpp [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#ifndef ANDROID
```
Comment (michalsek) [inline cpp/StereoPannerNode/ios/StereoPannerNodeWrapper.cpp:1]: another unnecessaryy `.keep` file left :)

File: ios/nodes/AudioNode/AudioNode.m [hunk 6]
```diff
@@ -1,29 +1,64 @@
...
+- (void)deprocess:(AVAudioPCMBuffer *)buffer playerNode:(AVAudioPlayerNode *)playerNode nodeToDeprocess:(AudioNode *)node {
+    for (AudioNode *node in _connectedNodes) {
+        [node deprocess:buffer playerNode:playerNode nodeToDeprocess:node];
+    }
+}
+
 - (void)connect:(AudioNode *)node {
-    if (self.numberOfOutputs > 0) {
-        [self.connectedNodes addObject:node];
+    if (_numberOfOutputs > 0) {
```
Comment (michalsek) [inline ios/nodes/AudioNode/AudioNode.m:28]: when/where this value is incremented or decremented?

File: ios/nodes/AudioNode/AudioNode.m [hunk 7]
```diff
@@ -1,29 +1,64 @@
...
-- (void)process {
-    for (AudioNode *node in self.connectedNodes) {
-        [node process];
+- (void)findNodesToDeprocess:(AudioNode *)node {
+    NSMutableArray<OscillatorNode *> *connectedNodes = _context.connectedOscillators;
+    for (OscillatorNode *cn in connectedNodes) {
+        if ([self findNodesToDeprocessHelper:cn]) {
+            [node deprocess:cn.buffer playerNode:cn.playerNode nodeToDeprocess:node];
+        }
+    }
```
Comment (michalsek) [inline ios/nodes/AudioNode/AudioNode.m]:
```suggestion
    NSMutableArray<OscillatorNode *> *connectedOscillators = _context.connectedOscillators;
    for (OscillatorNode *osNode in connectedOscillators) {
        if ([self findNodesToDeprocessHelper:osNode]) {
            [node deprocess:osNode.buffer playerNode:osNode.playerNode nodeToDeprocess:node];
        }
    }
```

File: ios/nodes/AudioNode/AudioNode.m (hunk 6 above)
Comment (hubgan) [reply ios/nodes/AudioNode/AudioNode.m:28]: https://developer.mozilla.org/en-US/docs/Web/API/AudioNode/numberOfInputs

This is little stupid but it's never decremented or incremented. This is hardcoded for specific nodes. For example Oscillator have numberOfOutputs equal to 1 but DestinationNode have numberOfOutputs equal to 0. If it's 0 we can connect anything to that node. Documentation isn't clear about it in my opinion. Below is an example written in Web Audio API.

```
const playSound = () => {
    if (!audioContextRef.current) {
      return;
    }

    const context = audioContextRef.current;
    const oscillator = context.createOscillator();
    console.log(oscillator.numberOfOutputs); // 1
    console.log(oscillator.numberOfInputs); // 0
    const gain = context.createGain();

    const time = context.currentTime;

    oscillator.frequency.value = FREQUENCY;
    gain.gain.setValueAtTime(1, time);
    gain.gain.linearRampToValueAtTime(
      FADE_OUT_GAIN_ATTACK,
      time + FADE_OUT_ATTACK_TIME
    );
    gain.gain.linearRampToValueAtTime(
      FADE_OUT_GAIN_RELEASE,
      time + FADE_OUT_RELEASE_TIME
    );

    oscillator.connect(gain).connect(context.destination);
    console.log(oscillator.numberOfOutputs); // 1
    console.log(oscillator.numberOfInputs); // 0
    oscillator.start(time);
    oscillator.stop(time + 0.1);
  };
```

File: cpp/GainNode/ios/GainNodeWrapper.cpp (hunk 4 above)
Comment (hubgan) [reply cpp/GainNode/ios/GainNodeWrapper.cpp:1]: I don't know why inside files changed it shows that .keep is an empty file. Every .keep file has been deleted, I checked on branch to be sure

File: cpp/AudioContext/ios/AudioContextWrapper.cpp [hunk 8]
```diff
@@ -1,14 +1,33 @@
...
+namespace audiocontext
+{
     std::shared_ptr<OscillatorNodeWrapper> AudioContextWrapper::createOscillator() {
-        return std::make_shared<OscillatorNodeWrapper>();
+        return std::make_shared<OscillatorNodeWrapper>(audiocontext_);
     }
 
     std::shared_ptr<AudioDestinationNodeWrapper> AudioContextWrapper::getDestination() {
-        throw std::runtime_error("[AudioContextHostObject::getDestination] Not yet implemented!");
+        // TODO: Add AudioDestinationNode implementation
```
Comment (michalsek) [inline cpp/AudioContext/ios/AudioContextWrapper.cpp:11]: Create an issue instead of a comment? 😛

**Pull request merged**